curifactory/staging.py

```diff
@@ -584,28 +669,81 @@ def wrapper(record: Record, records: list[Record] = None, **kwargs):
+
+            # if we have an execution list from our stage DAG, use that
+            # to determine if this stage executes or not.
+            if record.manager.map is not None:
+                stage_rep = (record.get_record_index(), name)
+                if stage_rep not in record.manager.map.execution_list:
+                    logging.debug('DAG-indicated stage skip "%s".' % str(stage_rep))
+                    _dag_skip_check_cached_outputs(
+                        name, record, outputs, cachers, records
+                    )
+
+                    # grab any possible previous reportables so they still end up in report.
+                    _check_cached_reportables(name, record)
+                    record.store_tracked_paths()
+                    execute_stage = False
+                else:
+                    # the representation is in the execution list, so execute!
+                    execute_stage = True
+
+            # otherwise, proceed normally with cache and load checks
+            # NOTE: this is an explicit _separate_ check because if our DAG indicates that
+            # this stage executes, we still want to actually double check the cache and determine
+            # if we _still_ need to execute this stage. This is primarily for cases where a
+            # stage happens to get run multiple times with different records that have the same args/
+            # same outputs. The DAG doesn't dynamically update with cached values during the actual
+            # experiment run, so it won't catch this case by itself.
+            # So the flow is: if we have a DAG, and it says to execute this stage, or if we don't have
+            # a DAG, check if we actually need to run this based on cached values.
+            if record.manager.map is None or execute_stage:
+                cache_valid = _check_cached_outputs(
+                    name, record, outputs, cachers, records
+                )
+                if cache_valid:
+                    # get previous reportables if available
+                    _check_cached_reportables(name, record, records)
+
+                    # if we've hit this point, we will be returning early/not executing
+                    # the stage because all outputs are found. The process of checking
+                    # cached outputs should correctly add all the necessary tracked paths,
+                    # so to transfer these paths into a store full run, we just need
+                    # the below call
+                    # NOTE: I _believe_ that we also get metadata from this because
+                    # the load_metadata will have entered the metadata path already.
+                    # this will need to be tested
+                    record.store_tracked_paths()
+                    execute_stage = False
+                else:
+                    # at least one output wasn't cached, so execute order 66!
+                    execute_stage = True

```

Comment:
This seems to be duplicate code as in `stage()`. Could refactor into its own function maybe?

Reply:
This is probably worth looking into in a later issue/PR. I ran into issues trying to abstract out certain types of functionality that were common between stage and aggregate:

https://github.com/ORNL/curifactory/blob/824f8a4c3a492f5977f5405b710f9cb11649a11a/curifactory/staging.py#L273-L276

and I'm not fully certain what makes that an issue, but it's something to be careful of.